src/Faker/Provider/Base.php

```diff
@@ -360,17 +375,21 @@ public static function numerify($string = '###')
      */
     public static function lexify($string = '????')
     {
-        return preg_replace_callback('/\?/u', 'static::randomLetter', $string);
+        return self::replaceWildcard($string, '?', 'static::randomLetter');
     }
 
     /**
-     * Replaces hash signs and question marks with random numbers and letters
+     * Replaces hash signs ('#') and question marks ('?') with random numbers and letters
+     * An asterisk ('*') is replaced with either a random number or a random letter
      *
      * @param  string $string String that needs to bet parsed
      * @return string
      */
     public static function bothify($string = '## ??')
     {
+        $string = self::replaceWildcard($string, '*', function () {
+            return mt_rand(0, 1) ? '#' : '?';
```

Comment:
How does it save a function call? For `randomLetter()`, it will be called once anyway and `randomDigit()` is not called at all, there's an optimization already that gets many digits from a big number.